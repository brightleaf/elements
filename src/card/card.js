import React from 'react'
import classnames from 'classnames'
import { Colors, Screens } from '../modifiers'

export const CardHeader = ({ title, hasIcon, fa, fas }) => {
  const classes = {
    fa,
    fas,
  }
  return (
    <header className="card-header">
      <p className="card-header-title">{title}</p>
      {hasIcon && (
        <a href="#" className="card-header-icon" aria-label="more options">
          <span className="icon">
            <i
              className={classnames(classes, 'fa-angle-down')}
              aria-hidden="true"
            ></i>
          </span>
        </a>
      )}
    </header>
  )
}
CardHeader.sortOrder = 0

export const CardFooterItem = ({ className, children }) => {
  const kid = React.cloneElement(children, {
    className: classnames('card-footer-item', className),
  })
  return <>{kid}</>
}
export const CardFooter = ({ children }) => {
  return <footer className="card-footer">{children}</footer>
}
CardFooter.sortOrder = 5

export const CardImage = ({ src, alt }) => {
  return (
    <div className="card-image">
      <figure className="image is-4by3">
        <img src={src} alt={alt} />
      </figure>
    </div>
  )
}
CardImage.sortOrder = 1
export const CardImageContainer = ({
  children,
  className,
  hasTextCentered,
}) => {
  const classes = {
    'has-text-centered': hasTextCentered,
  }
  return (
    <div className={classnames('card-image', className, classes)}>
      {children}
    </div>
  )
}
CardImageContainer.sortOrder = 1
export const CardBody = ({ children }) => {
  return <div className="card-content">{children}</div>
}
CardBody.sortOrder = 2

const CardX = ({
  children,
  isMobile,
  className,
  isFluid,
  isWideScreen,
  isFullHD,
}) => {
  const classes = Screens({ isMobile, isFluid, isWideScreen, isFullHD })
  return (
    <div className={classnames('card', className, classes)}>
      <header className="card-header">
        <p className="card-header-title">Component</p>
        <a href="#" className="card-header-icon" aria-label="more options">
          <span className="icon">
            <i className="fas fa-angle-down" aria-hidden="true"></i>
          </span>
        </a>
      </header>
      <div className="card-content">
        <div className="content">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus nec
          iaculis mauris.
          <a href="#">@bulmaio</a>. <a href="#">#css</a>{' '}
          <a href="#">#responsive</a>
          <br />
          <time dateTime="2016-1-1">11:09 PM - 1 Jan 2016</time>
        </div>
      </div>
      <footer className="card-footer">
        <a href="#" className="card-footer-item">
          Save
        </a>
        <a href="#" className="card-footer-item">
          Edit
        </a>
        <a href="#" className="card-footer-item">
          Delete
        </a>
      </footer>
    </div>
  )
}

export const Card = ({
  children,
  isMobile,
  isFluid,
  isWideScreen,
  isFullHD,
  className,
}) => {
  const classes = Screens({ isMobile, isFluid, isWideScreen, isFullHD })
  const childrenAsArray = React.Children.toArray(children).sort(
    (childA, childB) => {
      if (childA.type.sortOrder < childB.type.sortOrder) {
        return -1
      }
      if (childA.type.sortOrder > childB.type.sortOrder) {
        return 1
      }
      return 0
    }
  )
  var types = [CardHeader, CardBody, CardFooter, CardImage, CardImageContainer]
  childrenAsArray.forEach(child => {
    if (
      types.indexOf(child.type) > -1 ||
      (child.props.className && child.props.className.indexOf('card') > -1)
    ) {
      return
    }
    throw new Error(`'${child.type}' not allowed`)
  })
  return (
    <div className={classnames('card', className, classes)}>
      {childrenAsArray}
    </div>
  )
}

Card.propTypes = {
  children: function(props, propName, componentName) {
    const prop = props[propName]
    const types = [CardHeader, CardBody, CardImage, Card, CardImageContainer]
    // Only accept a single child, of the appropriate type
    if (propName === 'children') {
      if (!prop.forEach) {
        return
      }
      prop.forEach(p => {
        if (types.indexOf(p.type) === -1) {
          return new Error(`'${prop.type}' not allowed`)
        }
      })
    }
  },
}
Card.defaultProps = {
  className: '',
  isMobile: false,
  isFluid: false,
  isWideScreen: false,
  isFullHD: false,
}
