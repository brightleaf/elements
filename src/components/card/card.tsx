import React, { HTMLAttributes, FC } from "react";
import classnames from "classnames";
import { Screens } from "../../utils/modifiers";
import { allTheClasses } from "../../utils/modifiers";
import { BulmaProps } from "types";

export interface CardBaseProps<T> extends BulmaProps<T> {
  className?: string;
}

export interface CardHeaderIconProps extends HTMLAttributes<HTMLOrSVGElement> {
  className?: string;
  children?: React.ReactNode;
  onClick?: (e: React.MouseEvent<HTMLAnchorElement>) => void;
}

export const CardHeaderIcon: FC<CardBaseProps<HTMLAnchorElement>> = ({
  className,
  children,
  onClick,
  ...props
}) => {
  return (
    <a
      href="#"
      className={classnames(
        "card-header-icon",
        className,
        allTheClasses(props)
      )}
      aria-label="more options"
      onClick={onClick}
    >
      {children}
    </a>
  );
};
export interface CardHeaderProps<T> extends CardBaseProps<T> {
  title: string;
  hasIcon?: boolean;
}

export const CardHeader: FC<CardHeaderProps<HTMLElement>> = ({
  title,
  className,
  children,
  ...props
}) => {
  return (
    <header
      className={classnames("card-header", className, allTheClasses(props))}
    >
      <p className="card-header-title">{title}</p>
      {children}
    </header>
  );
};

export const CardFooterItem: FC<CardBaseProps<HTMLElement>> = ({
  className,
  children,
  ...props
}) => {
  const kid = React.Children.toArray(children).map((child) =>
    React.isValidElement(child)
      ? React.cloneElement(child, {
          className: classnames(
            "card-footer-item",
            className,
            allTheClasses(props)
          ),
        })
      : child
  );

  return <>{kid}</>;
};
export const CardFooter: FC<CardBaseProps<HTMLElement>> = ({
  children,
  className,
  ...props
}) => {
  return (
    <footer
      className={classnames("card-footer", className, allTheClasses(props))}
    >
      {children}
    </footer>
  );
};

export interface CardImageProps {
  className?: string;
  src: string;
  alt?: string;
}
export const CardImage: FC<CardImageProps> = ({ src, alt, className }) => {
  return (
    <div className={classnames("card-image", className)}>
      <figure className="image is-4by3">
        <img src={src} alt={alt} />
      </figure>
    </div>
  );
};

export interface CardImageContainerProps<T> extends CardBaseProps<T> {}
export const CardImageContainer: FC<CardImageContainerProps<
  HTMLDivElement
>> = ({ children, className, hasTextCentered, ...props }) => {
  const classes = allTheClasses(props);
  return (
    <div className={classnames("card-image", className, classes)}>
      {children}
    </div>
  );
};

export const CardBody: FC<CardBaseProps<HTMLDivElement>> = ({
  children,
  className,
  ...props
}) => {
  const classes = allTheClasses(props);
  return (
    <div className={classnames("card-content", className, classes)}>
      {children}
    </div>
  );
};

export const Card: FC<CardBaseProps<HTMLDivElement>> = ({
  children,
  className,
  onClick,
  ...props
}) => {
  const classes = allTheClasses(props);
  return (
    <div className={classnames("card", className, classes)} onClick={onClick}>
      {children}
    </div>
  );
};
