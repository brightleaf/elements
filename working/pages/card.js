/* eslint-disable spellcheck/spell-checker */
import React from 'react'
import {
  Card,
  CardBody,
  CardHeader,
  CardFooter,
  CardFooterItem,
  CardImage,
  Container,
  Column,
  Columns,
} from '../../src'

export default () => {
  return (
    <Container>
      <Columns>
        <Column isOneQuarter>
          <Card>
            <CardHeader title="Component" hasIcon></CardHeader>
            <CardBody>
              <div className="content">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                Phasellus nec iaculis mauris.
                <a href="#">@bulmaio</a>. <a href="#">#css</a>{' '}
                <a href="#">#responsive</a>
                <br />
                <time dateTime="2016-1-1">11:09 PM - 1 Jan 2016</time>
              </div>
            </CardBody>
            <CardFooter>
              {' '}
              <CardFooterItem>
                <a href="#">Save</a>
              </CardFooterItem>
              <a href="#" className="card-footer-item">
                Edit
              </a>
              <a href="#" className="card-footer-item">
                Delete
              </a>
            </CardFooter>
          </Card>
        </Column>
      </Columns>
      <hr />
      <Columns>
        <Column isOneQuarter>
          <Card>
            <CardImage
              src="https://bulma.io/images/placeholders/1280x960.png"
              alt="Placeholder image"
            />
            <CardBody>
              <p className="title">
                {
                  '"There are two hard things in computer science: cache invalidation, naming things, and off-by-one errors."'
                }
              </p>
              <p className="subtitle">Jeff Atwood</p>
            </CardBody>
            <CardFooter>
              <CardFooterItem>
                <p>
                  <span>
                    View on
                    <a href="https://twitter.com/codinghorror/status/506010907021828096">
                      Twitter
                    </a>
                  </span>
                </p>
              </CardFooterItem>
              <CardFooterItem>
                <p>
                  <span>
                    Share on <a href="#">Facebook</a>
                  </span>
                </p>
              </CardFooterItem>
            </CardFooter>
          </Card>
        </Column>
      </Columns>
    </Container>
  )
}
