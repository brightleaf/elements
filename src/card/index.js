import { Card, CardBody, CardHeader, CardFooter, CardImage } from './card'

export { Card, CardBody, CardHeader, CardFooter, CardImage }

export default { Card, CardBody, CardHeader, CardFooter, CardImage }
