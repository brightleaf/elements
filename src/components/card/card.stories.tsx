import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { Card, CardBody, CardHeader, CardHeaderIcon, CardFooter, CardFooterItem } from './card'
// import { Content } from '../content'


export default {
  title: 'Components/Card',
  component: Card,
} as ComponentMeta<typeof Card>

const Template: ComponentStory<typeof Card> = (args) => (
  <Card>
    <CardHeader title="Component" hasIcon></CardHeader>
    <CardBody>
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus nec iaculis mauris.
        <a href="#">@bulmaio</a>. <a href="#">#css</a> <a href="#">#responsive</a>
        <br />
        <time dateTime="2016-1-1">11:09 PM - 1 Jan 2016</time>
      </div>
    </CardBody>
    <CardFooter>
      <CardFooterItem>
        <a
          href="#"
          onClick={(e) => {
            console.info('clicked')
            e.preventDefault()
          }}
        >
          Save
        </a>
      </CardFooterItem>
      <CardFooterItem>
        <a href="#">Edit</a>
      </CardFooterItem>
      <CardFooterItem>
        <a href="#">Delete</a>
      </CardFooterItem>
    </CardFooter>
  </Card>
)

export const FullCardExample = Template.bind({})

FullCardExample.args = {

}
