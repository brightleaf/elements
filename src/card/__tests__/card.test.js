import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { toHaveClass } from '@testing-library/jest-dom'
import { Card, CardBody, CardHeader, CardFooter, CardImage } from '../card'

expect.extend({ toHaveClass })

describe('Card and components', () => {
  afterEach(cleanup)
  it('should render', () => {
    const { container } = render(
      <Card>
        <CardBody>Hi</CardBody>
      </Card>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render with header and footer and body', () => {
    const { container } = render(
      <Card>
        <CardHeader title={'HEADER'}></CardHeader>
        <CardBody>Hi</CardBody>
        <CardFooter title={'HEADER'}></CardFooter>
      </Card>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with footer header and body', () => {
    const { container } = render(
      <Card>
        <CardFooter></CardFooter>
        <CardHeader title={'HEADER'}></CardHeader>
        <CardBody>Hi</CardBody>
      </Card>
    )
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild.firstChild).toHaveClass('card-header')
    expect(container.firstChild.lastChild).toHaveClass('card-footer')
  })
})
