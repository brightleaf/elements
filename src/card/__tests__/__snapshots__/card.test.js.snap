// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card and components should render 1`] = `
<div
  class="card"
>
  <div
    class="card-content"
  >
    Hi
  </div>
</div>
`;

exports[`Card and components should render with footer header and body 1`] = `
<div
  class="card"
>
  <header
    class="card-header"
  >
    <p
      class="card-header-title"
    >
      HEADER
    </p>
  </header>
  <div
    class="card-content"
  >
    Hi
  </div>
  <footer
    class="card-footer"
  />
</div>
`;

exports[`Card and components should render with header and footer and body 1`] = `
<div
  class="card"
>
  <header
    class="card-header"
  >
    <p
      class="card-header-title"
    >
      HEADER
    </p>
  </header>
  <div
    class="card-content"
  >
    Hi
  </div>
  <footer
    class="card-footer"
  />
</div>
`;
